crypto/x509/x_x509.c

```diff
@@ -105,14 +105,16 @@ static int x509_cb(int operation, ASN1_VALUE **pval, const ASN1_ITEM *it,
             if (!ossl_x509_set0_libctx(ret, old->libctx, old->propq))
                 return 0;
             if (old->cert_info.key != NULL) {
-                EVP_PKEY *pkey = X509_PUBKEY_get0(old->cert_info.key);
+                EVP_PKEY *pkey_old = X509_PUBKEY_get(old->cert_info.key);
 
-                if (pkey != NULL) {
-                    pkey = EVP_PKEY_dup(pkey);
+                if (pkey_old != NULL) {
+                    EVP_PKEY *pkey = EVP_PKEY_dup(pkey_old);
                     if (pkey == NULL) {
-                        ERR_raise(ERR_LIB_X509, ERR_R_MALLOC_FAILURE);
-                        return 0;
-                    }
+                        pkey = pkey_old; // FIXME
```

Comment:
This is not right. You'd need to do at least EVP_PKEY_up_ref() here.

Reply:
`pkey` take ownership over `pkey_old` if `EVP_PKEY_dup` is failure. If we call `EVP_PKEY_up_ref()` here then associated keys reference counter will be equal 3 to the end of function. 
```c
            if (old->cert_info.key != NULL) { // old->cert_info.key._counter == 1
                EVP_PKEY *pkey_old = X509_PUBKEY_get(old->cert_info.key); // old->cert_info.key._counter == 2

                if (pkey_old != NULL) { // old->cert_info.key._counter == 2
                    EVP_PKEY *pkey = EVP_PKEY_dup(pkey_old);  // old->cert_info.key._counter == 2
                    if (pkey == NULL) { // old->cert_info.key._counter == 2
                        pkey = pkey_old; // FIXME  // old->cert_info.key._counter == 2
                    } else {
                        EVP_PKEY_free(pkey_old);  // old->cert_info.key._counter == 1
                    }

                    if (!X509_PUBKEY_set(&ret->cert_info.key, pkey)) {  // --old->cert_info.key._counter
                        EVP_PKEY_free(pkey); // --old->cert_info.key._counter
                        ERR_raise(ERR_LIB_X509, ERR_R_INTERNAL_ERROR);
                        return 0;
                    }
                    EVP_PKEY_free(pkey); // --old->cert_info.key._counter
```

Reply:
Ah, right, I did not notice the X509_PUBKEY_get0 -> get change

Reply:
Another, probably more valid, way would be to try to re-parse the public key bit string.

Reply:
Although I see X509_PUBKEY_dup doing the EVP_PKEY_up_ref() unconditionally. So perhaps we should do the same here for consistency and simply never do a _dup() here?

Reply:
> Another, probably more valid, way would be to try to re-parse the public key bit string.

Public key may contains not only bit string. For example group order for Epileptic Curve point. 

> So perhaps we should do the same here for consistency and simply never do a _dup() here?
My FIXME comment about that. I think that X509_dup  has to have own independent copy of public key. But my PR broke this. I think it's will be better to leave this as a bug, to support X509_dup for "ENGINE based" certs.